File: Fundamentals_of_Software_Design/2_Algorithms/Поиск подстроки в строке алгоритмом Рабина-Карпа/main.py
def my_strstr(haystack, needle):
    # Размеры строк
    n = len(haystack)
    m = len(needle)
    
    if m == 0:
        return 0  # Пустая подстрока всегда на первой позиции
    if n < m:
        return -1  # Если строка меньше подстроки, то совпадения нет
    
    # Параметры хеширования
    BASE = 256  # Основание для хеширования (число символов в алфавите, например, ASCII)
    MOD = 101   # Модуль для хеширования (простое число для уменьшения коллизий)
    
    # Вычисление хеша для строки
    def compute_hash(s, length):
        hash_val = 0
        for i in range(length):
            hash_val = (hash_val * BASE + ord(s[i])) % MOD
        return hash_val
    
    # Вычисление хеша для needle и первого окна в haystack
    needle_hash = compute_hash(needle, m)
    haystack_hash = compute_hash(haystack, m)
    
    # Предсчитываем степень BASE^(m-1) для использования в обновлении хеша
    base_m = 1
    for i in range(m - 1):
        base_m = (base_m * BASE) % MOD
    
    # Поиск подстроки
    for i in range(n - m + 1):
        # Если хеши совпали, проверяем символы
        if haystack_hash == needle_hash:
            if haystack[i:i + m] == needle:
                return i  # Совпадение найдено, возвращаем индекс
        # Если не на последней позиции, сдвигаем окно
        if i < n - m:
            haystack_hash = (haystack_hash - ord(haystack[i]) * base_m) % MOD
            haystack_hash = (haystack_hash * BASE + ord(haystack[i + m])) % MOD
    
    return -1  # Подстрока не найдена

# Пример тестирования
haystack = "this is a simple example"
needle = "simple"

result = my_strstr(haystack, needle)
if result != -1:
    print(f"Подстрока найдена на позиции: {result}")
else:
    print("Подстрока не найдена")

File: Fundamentals_of_Software_Design/2_Algorithms/Поиск подстроки в строке алгоритмом Рабина-Карпа/main.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define BASE 256       // Основание для хеширования (число символов в алфавите, например, ASCII)
#define MOD 101        // Модуль для хеширования (простое число, чтобы уменьшить коллизии)

// Вычисление хеша строки длины m
unsigned long long compute_hash(const char *str, int m) {
    unsigned long long hash = 0;
    for (int i = 0; i < m; i++) {
        hash = (hash * BASE + str[i]) % MOD;
    }
    return hash;
}

// Обновление хеша при сдвиге окна на один символ
unsigned long long roll_hash(unsigned long long old_hash, char old_char, char new_char, int m) {
    old_hash = (old_hash - old_char * power(BASE, m - 1)) % MOD; // Убираем старый символ
    old_hash = (old_hash * BASE + new_char) % MOD; // Добавляем новый символ
    return (old_hash + MOD) % MOD; // Модуль для предотвращения отрицательных значений
}

// Функция поиска подстроки алгоритмом Рабина-Карпа
int my_strstr(char *haystack, const char *needle) {
    int n = strlen(haystack);
    int m = strlen(needle);
    if (m == 0) return 0;  // Пустая строка всегда найдется на первой позиции
    if (n < m) return -1;  // Если строка меньше подстроки, то совпадения нет

    unsigned long long needle_hash = compute_hash(needle, m);  // Хеш подстроки
    unsigned long long haystack_hash = compute_hash(haystack, m);  // Хеш первого окна

    // Предсчитаем степень BASE^(m-1) для использования в roll_hash
    unsigned long long base_m = 1;
    for (int i = 0; i < m - 1; i++) {
        base_m = (base_m * BASE) % MOD;
    }

    // Смотрим на каждое окно длины m в строке haystack
    for (int i = 0; i <= n - m; i++) {
        // Если хеши совпали, проверяем совпадение строк
        if (haystack_hash == needle_hash) {
            if (strncmp(haystack + i, needle, m) == 0) {
                return i;  // Совпадение найдено
            }
        }

        // Если мы не достигли конца строки, сдвигаем окно
        if (i < n - m) {
            haystack_hash = roll_hash(haystack_hash, haystack[i], haystack[i + m], m);
        }
    }

    return -1;  // Подстрока не найдена
}

// Тестирование
int main() {
    char haystack[] = "this is a simple example";
    char needle[] = "simple";

    int result = my_strstr(haystack, needle);

    if (result != -1) {
        printf("Подстрока найдена на позиции: %d\n", result);
    } else {
        printf("Подстрока не найдена\n");
    }

    return 0;
}
